File: src-tauri/bindings/ts/src/transaction.ts
import { Transaction, TransactionInput, TransactionOutput, OutPoint } from './types';
import { BitQuanError } from './errors';
import { Wallet } from './wallet';

export class TransactionBuilder {
  private versionNum: number = 1;
  private inputs: TransactionInput[] = [];
  private outputs: TransactionOutput[] = [];
  private lockTimeVal: number = 0;
  private feeRateVal: number = 1000; // satoshis per byte

  version(version: number): TransactionBuilder {
    this.versionNum = version;
    return this;
  }

  addInput(previousTx: string, outputIndex: number): TransactionBuilder {
    const input: TransactionInput = {
      previous_output: {
        hash: previousTx,
        index: outputIndex
      },
      sequence: 0xFFFFFFFF
    };
    this.inputs.push(input);
    return this;
  }

  addOutput(address: string, amount: number): TransactionBuilder {
    const output: TransactionOutput = {
      value: amount,
      script_pubkey: this.addressToScriptPubkey(address),
      address
    };
    this.outputs.push(output);
    return this;
  }

  lockTime(lockTime: number): TransactionBuilder {
    this.lockTimeVal = lockTime;
    return this;
  }

  feeRate(feeRate: number): TransactionBuilder {
    this.feeRateVal = feeRate;
    return this;
  }

  build(): Transaction {
    if (this.inputs.length === 0) {
      throw new BitQuanError('INVALID_TRANSACTION', 'Transaction must have at least one input');
    }

    if (this.outputs.length === 0) {
      throw new BitQuanError('INVALID_TRANSACTION', 'Transaction must have at least one output');
    }

    const txSize = this.estimateSize();
    const fee = txSize * this.feeRateVal;

    return {
      version: this.versionNum,
      inputs: this.inputs,
      outputs: this.outputs,
      lock_time: this.lockTimeVal,
      fee,
      timestamp: Date.now()
    };
  }

  private estimateSize(): number {
    const baseSize = 10; // version + lock_time + input/output counts
    const inputSize = this.inputs.length * 40; // Approximate input size
    const outputSize = this.outputs.length * 34; // Approximate output size
    
    return baseSize + inputSize + outputSize;
  }

  private addressToScriptPubkey(address: string): string {
    // Simplified script pubkey generation
    return `OP_DUP OP_HASH160 ${address} OP_EQUALVERIFY OP_CHECKSIG`;
  }
}

export class TransactionUtils {
  static hash(transaction: Transaction): string {
    const serialized = this.serialize(transaction);
    return this.hashData(serialized);
  }

  static size(transaction: Transaction): number {
    return this.serialize(transaction).length;
  }

  static inputValue(transaction: Transaction): number {
    // In a real implementation, this would look up UTXO values
    return 0;
  }

  static outputValue(transaction: Transaction): number {
    return transaction.outputs.reduce((sum, output) => sum + output.value, 0);
  }

  static fee(transaction: Transaction): number {
    return transaction.fee;
  }

  static isValid(transaction: Transaction): boolean {
    return transaction.inputs.length > 0 && 
           transaction.outputs.length > 0 && 
           this.outputValue(transaction) > 0;
  }

  static serialize(transaction: Transaction): string {
    return JSON.stringify({
      version: transaction.version,
      inputs: transaction.inputs,
      outputs: transaction.outputs,
      lock_time: transaction.lock_time
    });
  }

  static deserialize(hex: string): Transaction {
    try {
      const data = JSON.parse(hex);
      return {
        version: data.version,
        inputs: data.inputs,
        outputs: data.outputs,
        lock_time: data.lock_time,
        fee: data.fee || 0,
        signature: data.signature,
        timestamp: data.timestamp
      };
    } catch (error) {
      throw new BitQuanError('DESERIALIZATION_ERROR', `Failed to deserialize transaction: ${error}`);
    }
  }

  static sign(transaction: Transaction, privateKey: string): Transaction {
    const signature = this.signTransaction(transaction, privateKey);
    return {
      ...transaction,
      signature
    };
  }

  static verifySignature(transaction: Transaction): boolean {
    return transaction.signature !== undefined;
  }

  private static hashData(data: string): string {
    // Simplified hashing - in production, use proper cryptographic hash functions
    let hash = 0;
    for (let i = 0; i < data.length; i++) {
      const char = data.charCodeAt(i);
      hash = ((hash << 5) - hash) + char;
      hash = hash & hash;
    }
    return Math.abs(hash).toString(16);
  }

  private static signTransaction(transaction: Transaction, privateKey: string): string {
    const serialized = this.serialize(transaction);
    const hash = this.hashData(serialized);
    return `signed_${privateKey.substring(0, 8)}_${hash}`;
  }
}

export class MerkleTree {
  private root: string;
  private leaves: string[];
  private depth: number;

  constructor(leaves: string[]) {
    this.leaves = [...leaves];
    this.depth = 0;
    this.root = this.build();
  }

  getRoot(): string {
    return this.root;
  }

  getProof(leafIndex: number): string[] {
    if (leafIndex >= this.leaves.length) {
      return [];
    }

    const proof: string[] = [];
    let currentLevel = [...this.leaves];
    let index = leafIndex;

    while (currentLevel.length > 1) {
      const isRight = index % 2 === 1;
      const siblingIndex = isRight ? index - 1 : index + 1;
      
      if (siblingIndex < currentLevel.length) {
        proof.push(currentLevel[siblingIndex]);
      } else {
        proof.push(currentLevel[index]);
      }

      const nextLevel: string[] = [];
      for (let i = 0; i < currentLevel.length; i += 2) {
        const left = currentLevel[i];
        const right = i + 1 < currentLevel.length ? currentLevel[i + 1] : left;
        const combined = left + right;
        nextLevel.push(this.hash(combined));
      }
      
      currentLevel = nextLevel;
      index = Math.floor(index / 2);
    }

    return proof;
  }

  static verifyProof(leaf: string, proof: string[], root: string): boolean {
    let currentHash = leaf;
    
    for (const sibling of proof) {
      const combined = currentHash + sibling;
      currentHash = MerkleTree.prototype.hash(combined);
    }
    
    return currentHash === root;
  }

  private build(): string {
    if (this.leaves.length === 0) {
      return '';
    }

    let currentLevel = [...this.leaves];
    this.depth = 0;

    while (currentLevel.length > 1) {
      const nextLevel: string[] = [];
      
      for (let i = 0; i < currentLevel.length; i += 2) {
        const left = currentLevel[i];
        const right = i + 1 < currentLevel.length ? currentLevel[i + 1] : left;
        const combined = left + right;
        nextLevel.push(this.hash(combined));
      }
      
      currentLevel = nextLevel;
      this.depth++;
    }

    return currentLevel[0];
  }

  private hash(data: string): string {
    // Simplified hashing - in production, use proper cryptographic hash functions
    let hash = 0;
    for (let i = 0; i < data.length; i++) {
      const char = data.charCodeAt(i);
      hash = ((hash << 5) - hash) + char;
      hash = hash & hash;
    }
    return Math.abs(hash).toString(16);
  }
}

File: src-tauri/crates/bitquan-sdk/src/transaction.rs
use crate::error::{BitQuanError, Result};
use crate::types::TransactionStatus;
use serde::{Deserialize, Serialize};
use sha2::Digest;


/// BitQuan transaction
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Transaction {
    pub version: u32,
    pub inputs: Vec<TransactionInput>,
    pub outputs: Vec<TransactionOutput>,
    pub lock_time: u64,
    pub fee: u64,
    pub signature: Option<String>,
    pub timestamp: Option<u64>,
}

/// Transaction input
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TransactionInput {
    pub previous_output: OutPoint,
    pub script_sig: Option<String>,
    pub sequence: u32,
}

/// Transaction output
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TransactionOutput {
    pub value: u64,
    pub script_pubkey: String,
    pub address: Option<String>,
}

/// Out point references a previous transaction output
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OutPoint {
    pub hash: String,
    pub index: u32,
}

/// Transaction builder for creating transactions
#[derive(Debug, Clone)]
pub struct TransactionBuilder {
    version: u32,
    inputs: Vec<TransactionInput>,
    outputs: Vec<TransactionOutput>,
    lock_time: u64,
    fee_rate: u64,
}

impl TransactionBuilder {
    /// Create a new transaction builder
    pub fn new() -> Self {
        Self {
            version: 1,
            inputs: Vec::new(),
            outputs: Vec::new(),
            lock_time: 0,
            fee_rate: 1000, // Default fee rate in satoshis per byte
        }
    }

    /// Set transaction version
    pub fn version(mut self, version: u32) -> Self {
        self.version = version;
        self
    }

    /// Add an input to the transaction
    pub fn add_input(mut self, previous_tx: &str, output_index: u32) -> Self {
        let input = TransactionInput {
            previous_output: OutPoint {
                hash: previous_tx.to_string(),
                index: output_index,
            },
            script_sig: None,
            sequence: u32::MAX,
        };
        self.inputs.push(input);
        self
    }

    /// Add an output to the transaction
    pub fn add_output(mut self, address: &str, amount: u64) -> Self {
        let output = TransactionOutput {
            value: amount,
            script_pubkey: address_to_script_pubkey(address),
            address: Some(address.to_string()),
        };
        self.outputs.push(output);
        self
    }

    /// Set the lock time
    pub fn lock_time(mut self, lock_time: u64) -> Self {
        self.lock_time = lock_time;
        self
    }

    /// Set the fee rate (satoshis per byte)
    pub fn fee_rate(mut self, fee_rate: u64) -> Self {
        self.fee_rate = fee_rate;
        self
    }

    /// Build the transaction
    pub fn build(self) -> Result<Transaction> {
        if self.inputs.is_empty() {
            return Err(BitQuanError::InvalidTransaction(
                "Transaction must have at least one input".to_string(),
            ));
        }

        if self.outputs.is_empty() {
            return Err(BitQuanError::InvalidTransaction(
                "Transaction must have at least one output".to_string(),
            ));
        }

        let tx_size = self.estimate_size();
        let fee = tx_size * self.fee_rate;

        Ok(Transaction {
            version: self.version,
            inputs: self.inputs,
            outputs: self.outputs,
            lock_time: self.lock_time,
            fee,
            signature: None,
            timestamp: None,
        })
    }

    /// Estimate transaction size in bytes
    fn estimate_size(&self) -> u64 {
        // Basic size estimation
        let base_size = 10; // version + lock_time + input/output counts
        let input_size = self.inputs.len() * 40; // Approximate input size
        let output_size = self.outputs.len() * 34; // Approximate output size
        
        (base_size + input_size + output_size) as u64
    }
}

impl Default for TransactionBuilder {
    fn default() -> Self {
        Self::new()
    }
}

impl Transaction {
    /// Create a new transaction
    pub fn new(version: u32, inputs: Vec<TransactionInput>, outputs: Vec<TransactionOutput>) -> Self {
        Self {
            version,
            inputs,
            outputs,
            lock_time: 0,
            fee: 0,
            signature: None,
            timestamp: None,
        }
    }

    /// Get transaction hash
    pub fn hash(&self) -> String {
        let serialized = self.serialize();
        hex::encode(sha2::Sha256::new().chain_update(serialized.as_bytes()).finalize())
    }

    /// Get transaction size in bytes
    pub fn size(&self) -> u64 {
        self.serialize().len() as u64
    }

    /// Get total input value
    pub fn input_value(&self) -> u64 {
        self.inputs.iter().map(|_| 0).sum() // Would need UTXO lookup
    }

    /// Get total output value
    pub fn output_value(&self) -> u64 {
        self.outputs.iter().map(|output| output.value).sum()
    }

    /// Get transaction fee
    pub fn fee(&self) -> u64 {
        self.fee
    }

    /// Check if transaction is valid
    pub fn is_valid(&self) -> bool {
        !self.inputs.is_empty() 
            && !self.outputs.is_empty()
            && self.output_value() > 0
    }

    /// Serialize transaction to hex string
    pub fn serialize(&self) -> String {
        // Simplified serialization
        format!(
            "v:{}|i:{}|o:{}|lt:{}",
            self.version,
            self.inputs.len(),
            self.outputs.len(),
            self.lock_time
        )
    }

    /// Deserialize transaction from hex string
    pub fn deserialize(_hex: &str) -> Result<Self> {
        // Simplified deserialization
        Err(BitQuanError::InvalidTransaction(
            "Deserialization not implemented".to_string(),
        ))
    }

    /// Sign transaction with private key
    pub fn sign(&mut self, _private_key: &str) -> Result<()> {
        // Simplified signing
        self.signature = Some("signed".to_string());
        Ok(())
    }

    /// Verify transaction signature
    pub fn verify_signature(&self) -> bool {
        self.signature.is_some()
    }
}

/// Convert address to script pubkey
fn address_to_script_pubkey(address: &str) -> String {
    // Simplified address to script conversion
    format!("OP_DUP OP_HASH160 {} OP_EQUALVERIFY OP_CHECKSIG", address)
}

/// Transaction pool information
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TransactionPoolInfo {
    pub size: usize,
    pub bytes: usize,
    pub usage: usize,
    pub max_size: usize,
}

/// Transaction receipt
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TransactionReceipt {
    pub transaction_hash: String,
    pub block_hash: Option<String>,
    pub block_number: Option<u64>,
    pub transaction_index: Option<u32>,
    pub status: TransactionStatus,
    pub gas_used: Option<u64>,
    pub cumulative_gas_used: Option<u64>,
    pub logs: Vec<TransactionLog>,
}

/// Transaction log
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TransactionLog {
    pub address: String,
    pub topics: Vec<String>,
    pub data: String,
    pub block_number: Option<u64>,
    pub transaction_hash: Option<String>,
    pub transaction_index: Option<u32>,
    pub block_hash: Option<String>,
    pub log_index: Option<u32>,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_transaction_builder() {
        let tx = TransactionBuilder::new()
            .add_input("prev_tx_hash", 0)
            .add_output("address1", 1000)
            .add_output("address2", 2000)
            .build()
            .unwrap();

        assert_eq!(tx.inputs.len(), 1);
        assert_eq!(tx.outputs.len(), 2);
        assert_eq!(tx.output_value(), 3000);
        assert!(tx.is_valid());
    }

    #[test]
    fn test_empty_transaction() {
        let result = TransactionBuilder::new().build();
        assert!(result.is_err());
    }

    #[test]
    fn test_transaction_hash() {
        let tx = TransactionBuilder::new()
            .add_input("prev_tx_hash", 0)
            .add_output("address1", 1000)
            .build()
            .unwrap();

        let hash = tx.hash();
        assert!(!hash.is_empty());
    }
}